Increase HttpClientHandler.MaxConnectionsPerServer to 64 to improve PM UI performance in Visual Studio

kartheekp-ms opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/11923

Regression? Last working version:

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->

Thank you @mjolka for identifying this performance improvement and creating the linked issue. For context, [`HttpClientHandler.MaxConnectionsPerServer`](https://docs.microsoft.com/dotnet/api/system.net.http.httpclienthandler.maxconnectionsperserver?view=net-6.0) is by default `2` in .NET framework but the default value in .NET Core is `Int32.MaxValue`. Reference https://docs.microsoft.com/en-gb/archive/blogs/timomta/controlling-the-number-of-outgoing-connections-from-httpclient-net-core-or-full-framework. The performance improvements of Visual Studio PM UI mentioned in the linked issue description were great. Hence, in this PR I propose to increase `HttpClientHandler.MaxConnectionsPerServer` to `64` in `.NET Framework` code paths but don't change this property in `.NET` code paths because of the default value i.e., `Int32.MaxValue`. NuGet.exe on Windows [sets max connections limit per endpoint to 64](https://github.com/NuGet/NuGet.Client/blob/a56c42dbdd901c0c611eece400ce8565fffdfdf7/src/NuGet.Core/NuGet.Common/NetworkProtocolUtility.cs#L19). Hence, I used the same value `64` in this PR. The proposed changes also allow customers to tweak `HttpClientHandler.MaxConnectionsPerServer` by adding [`maxHttpRequestsPerSource`](https://github.com/NuGet/Home/issues/4538) in [NuGet.Config](https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file) file.

I asked `.NET Networking team` if there would be any issues upgrading `MaxConnectionsPerServer` value in .NET Framework code paths. The response received so far from the team seems to be positive https://github.com/orgs/dotnet/teams/ncl/discussions/1. Hence, I went ahead with the changes. Thanks to @drewgillies for taking the time to test a private build with this fix. He mentioned that this fix made a lot of difference to the usability and performance of PM UI in Visual Studio.

Step | With preinstalled PMUI | With PMUI installed from a private build with this fix
-- | -- | --
Navigate to PMUI for solution, landing on Installed tab | 35s | 10s
Change to Browse tab | 36s | 13s
Change back to Installed tab | 15s | 3s
Change to Updates tab | 27s | 6s

~I don't have numbers handy right now, but I noticed great performance improvements in Visual Studio solution restore scenario. I will run a quick test and post the results here.~

EDIT: I captured few metrics. This PR reduced [OrchardCore](https://github.com/OrchardCMS/OrchardCore/commit/c4e37bdd6f63f3a5acdb018e101436cc5fd90a1f) VS Solution restore by **10 seconds**.

- VS Solution restore time without this fix.

MaxConnectionsPerSever | MaxHttpRequestsPerSource | Time for Solution restore
-- | -- | --
2 (Default) | Not set | **45 seconds**
2  (Default)| 64 | 42 seconds
2  (Default)| 128 | 39 seconds

- VS Solution restore time with this fix.

MaxConnectionsPerSever | MaxHttpRequestsPerSource | Time for Solution restore
-- | -- | --
64 | Default (64) | **35 seconds**
128 | 128 | 34 seconds

- `dotnet restore (7.0.100-preview.6)` for OrchardCore solution in my local machine took `32 seconds`. This number is close to VS solution restore scenario for default `MaxConnectionsPerSever` proposed in this PR which is `64`.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [x] N/A

kartheekp-ms committed:
update maxrequestspersource to 64 by default for .NET framework code paths

kartheekp-ms committed:
format

kartheekp-ms committed:
change code

kartheekp-ms committed:
fix bug

kartheekp-ms committed:
fix fail tests

kartheekp-ms committed:
code refactoring

kartheekp-ms committed:
added tests

nkolev92 commented:
> I think this PR also fixes https://github.com/NuGet/Home/issues/11638 issue. I can't reproduce this issue by installing VSIX from this PR build.

Personally, I have no problem with you marking that issue as being fixed as well.

If anyone from @NuGet/nuget-client disagrees, they should chime in :)

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/HttpSource/HttpHandlerResourceV3Provider.cs:
```diff
@@ -48,6 +48,13 @@ private static HttpHandlerResourceV3 CreateResource(PackageSource packageSource)
                 AutomaticDecompression = (DecompressionMethods.GZip | DecompressionMethods.Deflate)
             };
 
+#if IS_DESKTOP
```
If people set the MaxHTTPRequestPerSource explicitly we should *always* respect that.

Doesn't this break that in dotnet.exe?

kartheekp-ms replied:
If customers have configured `MaxHttpRequestsPerSource` then throttling requests is controlled by following code in both `.NET` and `.NET Framework` code paths. we don't change the value to `64` (which is new default). 

https://github.com/NuGet/NuGet.Client/blob/a56c42dbdd901c0c611eece400ce8565fffdfdf7/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSourceResourceProvider.cs#L46-L49

The only difference is, in `.NET Framework` code path if customers configure `MaxHttpRequestsPerSource` then above logic controls throttling requests and we also set `HttpClientHandler.MaxConnectionsPerServer` to `MaxHttpRequestsPerSource` value. The reason is customers may expect NuGet to send that many requests to a source in parallel but at the end because of the default `MaxConnectionsPerServer` only `2` requests are sent. Whereas in `.NET` code paths above logic controls throttling requests and we don't `MaxConnectionsPerServer` because its default value is `Int32.MaxValue`.

zivkan commented:
> > I think this PR also fixes https://github.com/NuGet/Home/issues/11638 issue. I can't reproduce this issue by installing VSIX from this PR build.

> If anyone from https://github.com/orgs/NuGet/teams/nuget-client disagrees, they should chime in

I disagree. If VS was hanging when making network calls, that needs to be fixed by not wating for the network to finish on the UI thread. So, it might have been fixed by https://github.com/NuGet/NuGet.Client/pull/4739.  If it wasn't, then the duration of the UI delay might be reduced by this PR, but it doesn't fix network calls being made on the UI thread.

nkolev92 commented:
> > > I think this PR also fixes [NuGet/Home#11638](https://github.com/NuGet/Home/issues/11638) issue. I can't reproduce this issue by installing VSIX from this PR build.
> 
> > If anyone from https://github.com/orgs/NuGet/teams/nuget-client disagrees, they should chime in
> 
> I disagree. If VS was hanging when making network calls, that needs to be fixed by not wating for the network to finish on the UI thread. So, it might have been fixed by #4739. If it wasn't, then the duration of the UI delay might be reduced by this PR, but it doesn't fix network calls being made on the UI thread.

I actually agree with you :D I didn't read that issue in too much detail. It was actually a real hang, with UI freezes and all. 
This change simply can't fix that.

kartheekp-ms commented:
> > > > I think this PR also fixes [NuGet/Home#11638](https://github.com/NuGet/Home/issues/11638) issue. I can't reproduce this issue by installing VSIX from this PR build.
> > 
> > 
> > > If anyone from https://github.com/orgs/NuGet/teams/nuget-client disagrees, they should chime in
> > 
> > 
> > I disagree. If VS was hanging when making network calls, that needs to be fixed by not wating for the network to finish on the UI thread. So, it might have been fixed by #4739. If it wasn't, then the duration of the UI delay might be reduced by this PR, but it doesn't fix network calls being made on the UI thread.
> 
> I actually agree with you :D I didn't read that issue in too much detail. It was actually a real hang, with UI freezes and all. This change simply can't fix that.

@zivkan @nkolev92 I agree with both of you :). I am unable to reproduce this issue https://github.com/NuGet/Home/issues/11638 even on the internal dogfooding version. Given that @jebriede fixed this UI delay in https://github.com/NuGet/NuGet.Client/pull/4739, I would let him close https://github.com/NuGet/Home/issues/11638 this issue as well.

jebriede commented:
> > > > > I think this PR also fixes [NuGet/Home#11638](https://github.com/NuGet/Home/issues/11638) issue. I can't reproduce this issue by installing VSIX from this PR build.
> > > 
> > > 
> > > > If anyone from https://github.com/orgs/NuGet/teams/nuget-client disagrees, they should chime in
> > > 
> > > 
> > > I disagree. If VS was hanging when making network calls, that needs to be fixed by not wating for the network to finish on the UI thread. So, it might have been fixed by #4739. If it wasn't, then the duration of the UI delay might be reduced by this PR, but it doesn't fix network calls being made on the UI thread.
> > 
> > 
> > I actually agree with you :D I didn't read that issue in too much detail. It was actually a real hang, with UI freezes and all. This change simply can't fix that.
> 
> @zivkan @nkolev92 I agree with both of you :). I am unable to reproduce this issue [NuGet/Home#11638](https://github.com/NuGet/Home/issues/11638) even on the internal dogfooding version. Given that @jebriede fixed this UI delay in #4739, I would let him close [NuGet/Home#11638](https://github.com/NuGet/Home/issues/11638) this issue as well.

@kartheekp-ms The fix I made for #4739 fixed a specific code path from running on the UI thread that was rebuilding cache files on the file system. The fix was not expecting or attempting to resolve all issues where code paths originating from the PMUI were running on the UI thread because each such code path originates from different places. Any other occurrences will need to be addressed separately.

nkolev92 commented:
I think the problem in https://github.com/NuGet/Home/issues/11638 is that we don't understand what exact codepath no?
Like it could be the same root cause as the one fixed in #4739, but we probably can't know now.

kartheekp-ms committed:
updated test names

kartheekp-ms committed:
updated tests

kartheekp-ms committed:
made tests async

kartheekp-ms commented:
I noticed typos in the test names I added in this PR. Hence pushed a new commit updating the test names and added a comment to explain why discard is used to ignore a return value in one of the test cases.

zivkan commented on test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSourceResourceProviderTests.cs:
```diff
@@ -0,0 +1,63 @@
+        public async Task WhenMaxHttpRequestsPerSourceIsNotConfiguredThenItsValueIsSetToDefault_SuccessAsync()
+        {
+            // Arrange
+            var packageSource = new PackageSource("https://contoso.com/v3/index.json");
```
```suggestion
            var packageSource = new PackageSource("https://contoso.test/v3/index.json");
```

We should never use potentially real URLs in tests, since we don't want tests to accidentally send HTTP requests to real servers.

zivkan commented on test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSourceResourceProviderTests.cs:
```diff
@@ -0,0 +1,63 @@
+        public async Task WhenMaxHttpRequestsPerSourceIsNotConfiguredThenItsValueWillNotBeUpdated_SuccessAsync()
+        {
+            // Arrange
+            var packageSource = new PackageSource("https://contoso.com/v3/index.json");
```
```suggestion
            var packageSource = new PackageSource("https://contoso.test/v3/index.json");
```

zivkan commented on test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSourceResourceProviderTests.cs:
```diff
@@ -0,0 +1,63 @@
+        public async Task WhenMaxHttpRequestsPerSourceIsConfiguredThenItsValueWillNotBeUpdated_SuccessAsync(int maxHttpRequestsPerSource)
+        {
+            // Arrange
+            var packageSource = new PackageSource("https://contoso.com/v3/index.json") { MaxHttpRequestsPerSource = maxHttpRequestsPerSource };
```
```suggestion
            var packageSource = new PackageSource("https://contoso.test/v3/index.json") { MaxHttpRequestsPerSource = maxHttpRequestsPerSource };
```

zivkan commented on test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpHandlerResourceV3ProviderTests.cs:
```diff
@@ -0,0 +1,83 @@
+        public async Task PackageSourceMaxHttpRequestsPerSourceIsNotForwardedToV3HttpClientHandler_SuccessAsync(int maxHttpRequestsPerSource)
+        {
+            // Arrange
+            var packageSource = new PackageSource("https://contoso.com/v3/index.json") { MaxHttpRequestsPerSource = maxHttpRequestsPerSource };
```
```suggestion
            var packageSource = new PackageSource("https://contoso.test/v3/index.json") { MaxHttpRequestsPerSource = maxHttpRequestsPerSource };
```

zivkan commented on test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpHandlerResourceV3ProviderTests.cs:
```diff
@@ -0,0 +1,83 @@
+        public async Task PackageSourceMaxHttpRequestsPerSourceIsForwardedToV3HttpClientHandler_SuccessAsync(int maxHttpRequestsPerSource)
+        {
+            // Arrange
+            var packageSource = new PackageSource("https://contoso.com/v3/index.json") { MaxHttpRequestsPerSource = maxHttpRequestsPerSource };
```
```suggestion
            var packageSource = new PackageSource("https://contoso.test/v3/index.json") { MaxHttpRequestsPerSource = maxHttpRequestsPerSource };
```

zivkan commented on test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpHandlerResourceV3ProviderTests.cs:
```diff
@@ -0,0 +1,83 @@
+        public async Task DefaultMaxHttpRequestsPerSourceIsForwardedToV3HttpClientHandler_SuccessAsync()
+        {
+            // Arrange
+            var packageSource = new PackageSource("https://contoso.com/v3/index.json");
```
```suggestion
            var packageSource = new PackageSource("https://contoso.test/v3/index.json");
```

kartheekp-ms committed:
updated package source url

kartheekp-ms commented:
I have manually queued a PR build by updating the variable to run end to end/apex tests on VS 17.3. CI checks passed for this PR. I am planning to merge this PR today.
